app/src/main/java/org/schabi/newpipe/fragments/MainFragment.java

```diff
@@ -59,56 +90,145 @@
 
         tabLayout.setupWithViewPager(viewPager);
 
-        int channelIcon = ThemeHelper.resolveResourceIdFromAttr(activity, R.attr.ic_channel);
-        int whatsHotIcon = ThemeHelper.resolveResourceIdFromAttr(activity, R.attr.ic_hot);
-        int bookmarkIcon = ThemeHelper.resolveResourceIdFromAttr(activity, R.attr.ic_bookmark);
-
-        if (isSubscriptionsPageOnlySelected()) {
-            tabLayout.getTabAt(0).setIcon(channelIcon);
-            tabLayout.getTabAt(1).setIcon(bookmarkIcon);
-        } else {
-            tabLayout.getTabAt(0).setIcon(whatsHotIcon);
-            tabLayout.getTabAt(1).setIcon(channelIcon);
-            tabLayout.getTabAt(2).setIcon(bookmarkIcon);
+        mainPageChanged();
+    }
+
+
+    public void mainPageChanged() {
+        getTabOrder();
+        adapter.notifyDataSetChanged();
+        viewPager.setOffscreenPageLimit(adapter.getCount());
+        setIcons();
+        setFirstTitle();
+    }
+
+    private void setFirstTitle() {
+        if((tabs.size()>0)&&activity != null) {
+            String tabNumber = tabs.get(0);
+
+                if (tabNumber.startsWith(TAB_NUMBER_KIOSK + "\t")) {
+                    String kiosk[] = tabNumber.split("\t");
+                    if (kiosk.length == 2) {
+                        try {
+                            setTitle(kiosk[1]);
+                        } catch (Exception e) {
+                            //ignore this. It WILL be thrown while the service is changed.
+                        }
+                    }
+                } else if (tabNumber.startsWith(TAB_NUMBER_CHANNEL + "\t")) {
+
+                    String channelInfo[] = tabNumber.split("\t");
+                    if(channelInfo.length==4) {
+                        setTitle(channelInfo[2]);
+                    }
+                } else {
+                    switch (tabNumber) {
+                        case TAB_NUMBER_BLANK:
+                            setTitle(getString(R.string.app_name));
+                            break;
+                        case TAB_NUMBER_SUBSCIRPTIONS:
+                            setTitle(getString(R.string.tab_subscriptions));
+                            break;
+                        case TAB_NUMBER_FEED:
+                            setTitle(getString(R.string.fragment_whats_new));
+                            break;
+                        case TAB_NUMBER_BOOKMARKS:
+                            setTitle(getString(R.string.tab_bookmarks));
+                            break;
+                        case TAB_NUMBER_HISTORY:
+                            setTitle(getString(R.string.title_activity_history));
+                            break;
+                    }
+                }
+
+
         }
     }
 
-    /*//////////////////////////////////////////////////////////////////////////
-    // Menu
-    //////////////////////////////////////////////////////////////////////////*/
+    private void setIcons() {
```

Comment:
Cant the icons and the titles be setup simultaniously? This looks like redundant code. Maybe you should consider a structure where you call ´setupTabs()` where you don't have to write the switch case etc. twice.